### PR TITLE
Upgrade notification handling

### PR DESCRIPTION
**Pull request by we-ai**
This PR implements improved notification handling, as discussed in [issue#1052](https://github.com/episphere/connect/issues/1052). Below are details:
- Use http requests (generated by cloud scheduler) to trigger notification handler (`sendScheduledNotifications`).
- Deploy notification handler as 2nd generation cloud function, so that it can run up to 1 hour after an http trigger.
- Send out SMS messages in bulk.
- Handle bulk messages status updates from Twilio.

### CODE REVIEW

File: utils/notifications.js [hunk 1]
```diff
@@ -463,13 +492,21 @@ async function getParticipantsAndSendNotifications({ notificationSpec, cutoffTim
   
       if (smsRecordArray.length > 0) {
         try {
-          smsRecordArray = await sendSmsBatch(smsRecordArray);
-          if (smsRecordArray.length > 0) {
-            await saveNotificationBatch(smsRecordArray);
-            smsCount[lang] += smsRecordArray.length;
+          const smsNotification = await sendBulkSms(phoneNumberArray, smsInSpec[lang].body);
+          for (const smsRecord of smsRecordArray) {
+            smsRecord.sid = smsNotification.sid;
```
Comment (jhflorey) [inline utils/notifications.js:497]: i have not tried the code as your suggestion from the comment https://github.com/episphere/connect/issues/1052#issuecomment-2284263069, As my understanding all SMS message records will have same `sid` (it should be `messageSid`), and when we get result from webhook it seems very difficult to find out exact status of each SMS message

**Review by jhflorey — APPROVED** [review]

**Review by anthonypetersen — APPROVED** [review]
looks good

File: utils/notifications.js (hunk 1 above)
Comment (we-ai) [reply utils/notifications.js:497]: For single message sending(`twilioClient.messages.create()`), `sid` in response is `messageSid` in Twilio.
While for bulk messages in this case, `sid` in `smsNotification` is not the `messgeSid` in Twilio.
I suggest you check Twilio data and webhook response data, to see how to update delivery status of each message to its record Firestore. @jhflorey 
Can `serviceSid` be used in webhook?

Comment (jhflorey) [reply utils/notifications.js:497]: I am focusing on another issue. I will try your suggestion after I finish my work.

Comment (jhflorey) [reply utils/notifications.js:497]: @we-ai  i just tried sending bulk sms to "+14159098560", "+13214567890". the response is the data below
```

{
  sid: 'NTe9a6f25490500956b893ce7e82f25e0c',
  accountSid: 'AC065580740768346cee7b13596c145d89',
  serviceSid: 'ISf4d9d06bc0d68eefd982f3a0c2258264',
  dateCreated: 2024-08-15T17:37:23.000Z,
  identities: [],
  tags: [],
  segments: [],
  priority: 'high',
  ttl: 2419200,
  title: null,
  body: 'This is a test message',
  sound: null,
  action: null,
  data: null,
  apn: null,
  gcm: null,
  fcm: null,
  sms: null,
  facebookMessenger: null,
  alexa: null
}
```

Comment (jhflorey) [reply utils/notifications.js:497]: The data of webhook
```

{
  MessagingServiceSid: 'MGadfd720a874372077c4d504530ed227a',
  ApiVersion: '2010-04-01',
  MessageStatus: 'delivered',
  RawDlrDoneDate: '2408151737',
  SmsSid: 'SM901f7b137dd71201bdc3bc0abbdadd55',
  SmsStatus: 'delivered',
  To: '+13214567890',
  From: '+13018041146',
  MessageSid: 'SM901f7b137dd71201bdc3bc0abbdadd55',
  AccountSid: 'AC065580740768346cee7b13596c145d89'
}
{
  MessagingServiceSid: 'MGadfd720a874372077c4d504530ed227a',
  ApiVersion: '2010-04-01',
  MessageStatus: 'delivered',
  RawDlrDoneDate: '2408151737',
  SmsSid: 'SM759eeea12ebda40be52f1ebc84a7e24e',
  SmsStatus: 'delivered',
  To: '+14159098560',
  From: '+13018041146',
  MessageSid: 'SM759eeea12ebda40be52f1ebc84a7e24e',
  AccountSid: 'AC065580740768346cee7b13596c145d89'
}
```

Comment (jhflorey) [reply utils/notifications.js:497]: It looks like we can not match the webhook's data with the notification in firestore if using sendBulkSms

Comment (we-ai) [reply utils/notifications.js:497]: Above info were edited for security.

Comment (we-ai) [reply utils/notifications.js:497]: > It looks like we can not match the webhook's data with the notification in firestore if using sendBulkSms

Ok. I'm checking to see whether there're other ways to do it.

File: utils/firestore.js [hunk 2]
```diff
@@ -3341,6 +3340,25 @@ const getAppSettings = async (appName, selectedParamsArray) => {
...
 }
 
+/**
+ * Update Notify message delivery status to Firestore.
+ * @param {Object} data 
+ */
+const updateNotifySmsRecord = async (data) => {
+  const snapshot = await db
+    .collection("notifications")
+    .where("phone", "==", data.phone)
```
Comment (jhflorey) [inline utils/firestore.js:3350]: Is there any case where phone number will be duplicated?

File: utils/webhook.js [hunk 3]
```diff
@@ -63,8 +64,25 @@ const webhook = async (req, res) => {
     }
 
     const query = req.query;
+    if (query.api === "twilio-notify-callback" && req.body.IsFinal === "true") {
```
Comment (jhflorey) [inline utils/webhook.js:67]: could you share where did we define this webhook api in twilio?

Comment (we-ai) [reply utils/webhook.js:67]: Not in Twilio. It's in API call to Twilio.

File: utils/firestore.js (hunk 2 above)
Comment (we-ai) [reply utils/firestore.js:3350]: No. Phone numbers are de-duplicated, so that,  in an API call of bulk messages, each phone number is unique.

**Pull request merged**